File: 08/python/main.py
from typing import List, Optional
import functools
import itertools

segments_to_digit = {
    frozenset("abcefg"): 0,
    frozenset("cf"): 1,
    frozenset("acdeg"): 2,
    frozenset("acdfg"): 3,
    frozenset("bcdf"): 4,
    frozenset("abdfg"): 5,
    frozenset("abdefg"): 6,
    frozenset("acf"): 7,
    frozenset("abcdefg"): 8,
    frozenset("abcdfg"): 9,
}
digits = frozenset(range(10))
segment_names = ["a", "b", "c", "d", "e", "f", "g"]


def part1(entries):
    return sum(len(val) in [2, 3, 4, 7] for _, rhs in entries for val in rhs)


def part2(entries):
    def solve(signal_patterns, outputs):
        def decode(pattern, assignment) -> Optional[int]:
            segments = frozenset((assignment[c] for c in pattern))
            return segments_to_digit.get(segments)

        def is_valid(assignment):
            decoded_nums = {decode(pattern, assignment) for pattern in signal_patterns}
            return decoded_nums == digits

        def construct_num(digits: List[int]) -> int:
            return functools.reduce(lambda acc, d: acc * 10 + d, digits)

        assignments = (
            dict(zip(segment_names, ordering))
            for ordering in itertools.permutations(segment_names)
        )
        valid_assignment = next(filter(is_valid, assignments))
        output_digits = (decode(pattern, valid_assignment) for pattern in outputs)
        return construct_num(output_digits)

    return sum(solve(*entry) for entry in entries)


with open("../input.txt") as f:
    splits = [line.split("|") for line in f]
    entries = [(lhs.strip().split(), rhs.strip().split()) for lhs, rhs in splits]
    print(part1(entries))
    print(part2(entries))

File: 08/rust/src/main.rs
use crate::Segment::{A, B, C, D, E, F, G};
use enum_map::{enum_map, Enum, EnumMap};
use enumset::{enum_set, EnumSet, EnumSetType};
use itertools::Itertools;
use std::collections::HashMap;

#[macro_use]
extern crate lazy_static;

struct Entry {
    signal_patterns: [EnumSet<Segment>; 10],
    output_patterns: [EnumSet<Segment>; 4],
}

#[derive(Debug, Hash, EnumSetType, Enum)]
enum Segment {
    A,
    B,
    C,
    D,
    E,
    F,
    G,
}

type Display = EnumSet<Segment>;

lazy_static! {
    static ref DISPLAY_TO_DIGIT: HashMap<Display, usize> = HashMap::from([
        (enum_set!(A | B | C | E | F | G), 0),
        (enum_set!(C | F), 1),
        (enum_set!(A | C | D | E | G), 2),
        (enum_set!(A | C | D | F | G), 3),
        (enum_set!(B | C | D | F), 4),
        (enum_set!(A | B | D | F | G), 5),
        (enum_set!(A | B | D | E | F | G), 6),
        (enum_set!(A | C | F), 7),
        (enum_set!(A | B | C | D | E | F | G), 8),
        (enum_set!(A | B | C | D | F | G), 9),
    ]);
}

const UNIQUE_NUM_SEGMENTS: [usize; 4] = [1, 3, 4, 7];

fn main() -> anyhow::Result<()> {
    let input = include_str!("../../input.txt");
    let entries = input
        .lines()
        .map(parse_entry)
        .collect::<Result<Vec<Entry>, _>>()?;
    println!("part 1: {}", part1(&entries));
    println!("part 2: {}", part2(&entries));
    Ok(())
}

fn part1(entries: &[Entry]) -> usize {
    entries
        .iter()
        .flat_map(|entry| entry.output_patterns)
        .filter(|segments| UNIQUE_NUM_SEGMENTS.contains(&segments.len()))
        .count()
}

fn part2(entries: &[Entry]) -> usize {
    entries.iter().map(solve).sum()
}

fn solve(entry: &Entry) -> usize {
    let segments = [A, B, C, D, E, F, G];
    let assignment = segments
        .into_iter()
        .permutations(7)
        .map(|ordering| {
            let mut result: EnumMap<Segment, Segment> = enum_map! {_ => A};
            segments
                .into_iter()
                .zip(ordering)
                .for_each(|(l, r)| result[l] = r);
            result
        })
        .find(|assignment| valid_assignment(*assignment, entry.signal_patterns))
        .expect("No valid assignments");
    entry
        .output_patterns
        .iter()
        .map(|&display| transform_display(display, assignment))
        .map(read_digit)
        .fold(0, |acc, digit| {
            acc * 10 + digit.expect("invalid rhs display")
        })
}

fn valid_assignment(assignment: EnumMap<Segment, Segment>, signal_patterns: [Display; 10]) -> bool {
    signal_patterns
        .iter()
        .map(|pattern| transform_display(*pattern, assignment))
        .filter_map(read_digit)
        .unique()
        .count()
        == 10
}

fn transform_display(display: Display, mapping: EnumMap<Segment, Segment>) -> Display {
    display.iter().map(|segment| mapping[segment]).collect()
}

fn read_digit(display: Display) -> Option<usize> {
    DISPLAY_TO_DIGIT.get(&display).copied()
}

fn parse_entry(s: &str) -> anyhow::Result<Entry> {
    let (lhs, rhs) = s.trim().split('|').next_tuple().unwrap();
    let lhs: Vec<Display> = lhs
        .trim()
        .split_whitespace()
        .map(parse_display)
        .collect::<anyhow::Result<Vec<_>>>()?;
    let rhs = rhs
        .trim()
        .split_whitespace()
        .map(parse_display)
        .collect::<anyhow::Result<Vec<_>>>()?;
    Ok(Entry {
        signal_patterns: lhs.try_into().expect("wrong lhs size"),
        output_patterns: rhs.try_into().expect("wrong rhs size"),
    })
}

fn parse_display(s: &str) -> anyhow::Result<Display> {
    s.chars()
        .map(|c| match c {
            'a' => Ok(A),
            'b' => Ok(B),
            'c' => Ok(C),
            'd' => Ok(D),
            'e' => Ok(E),
            'f' => Ok(F),
            'g' => Ok(G),
            _ => Err(anyhow::format_err!("invalid character: {}", c)),
        })
        .collect::<anyhow::Result<Display>>()
}
